### PR TITLE
Added noncentral-chisquared, noncentral-f and logseries distributions

### PR DESCRIPTION
**Pull request by kc611**
As titled, this PR adds support for following distributions:

- `numpy.random.Generator.logseries`
- `numpy.random.Generator.noncentral_chisquare`
- `numpy.random.Generator.noncentral_f`

### CODE REVIEW

File: numba/np/random/generator_methods.py [hunk 1]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+        def impl(inst, df, nonc, size=None):
+            return np.float64(random_noncentral_chisquare(inst.bit_generator,
+                                                          df, nonc))
+        return impl
+    else:
+        check_size(size)
+
+        def impl(inst, df, nonc, size=None):
+            out = np.empty(size, dtype=np.float64)
+            for i in np.ndindex(size):
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:842]: This loop is going to be doing a lot of work anyway, but in general, `out` is going to be C ordered by nature of its construction so iterating through `out.ravel()` or `out.flat()` would probably generate a less involved loop nest.

File: numba/np/random/generator_methods.py [hunk 2]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
                 out[i] = random_negative_binomial(inst.bit_generator, n, p)
             return out
         return impl
+
+
+@overload_method(types.NumPyRandomGeneratorType, 'noncentral_chisquare')
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:825]: Need to add all these methods to the documentation.

File: numba/np/random/generator_methods.py [hunk 3]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+def NumPyRandomGeneratorType_noncentral_chisquare(inst, df, nonc, size=None):
+    check_types(df, [types.Float, float], 'df')
+    check_types(nonc, [types.Float, float], 'nonc')
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:828]: The NumPy docs suggest that `df` and `nonc` can be an `array_like` of floats too. Having the restriction above is fine for an initial implementation but should be documented.

File: numba/np/random/generator_methods.py [hunk 4]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+                                                     df, nonc)
+            return out
+        return impl
+
+
+@overload_method(types.NumPyRandomGeneratorType, 'noncentral_f')
+def NumPyRandomGeneratorType_noncentral_f(inst, dfnum, dfden, nonc, size=None):
+    check_types(dfnum, [types.Float, float], 'dfnum')
+    check_types(dfden, [types.Float, float], 'dfden')
+    check_types(nonc, [types.Float, float], 'nonc')
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:853]: Same comment RE accepting `array_like` of floats and things that will cast to `float`.

File: numba/np/random/generator_methods.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/np/random/generator_methods.py:828]: NumPy also seems to "work" for types that will cast to float, e.g. `generator.noncentral_chisquare(12, 7, size=(3, 4))` works fine.

File: numba/np/random/generator_methods.py [hunk 5]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
         return impl
+
+
+@overload_method(types.NumPyRandomGeneratorType, 'noncentral_chisquare')
+def NumPyRandomGeneratorType_noncentral_chisquare(inst, df, nonc, size=None):
+    check_types(df, [types.Float, float], 'df')
+    check_types(nonc, [types.Float, float], 'nonc')
+    if isinstance(size, types.Omitted):
+        size = size.value
+
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:831]: The implementations need to raise `ValueError` if `df <= 0` or `nonc < 0`.

File: numba/np/random/generator_methods.py [hunk 6]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+
+@overload_method(types.NumPyRandomGeneratorType, 'noncentral_f')
+def NumPyRandomGeneratorType_noncentral_f(inst, dfnum, dfden, nonc, size=None):
+    check_types(dfnum, [types.Float, float], 'dfnum')
+    check_types(dfden, [types.Float, float], 'dfden')
+    check_types(nonc, [types.Float, float], 'nonc')
+    if isinstance(size, types.Omitted):
+        size = size.value
+
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:856]: The implementations need to raise `ValueError` if `df <= 0`, `dfden <= 0` or `nonc < 0`.

File: numba/np/random/generator_methods.py [hunk 7]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+            for i in np.ndindex(size):
+                out[i] = random_noncentral_f(inst.bit_generator,
+                                             dfnum, dfden, nonc)
+            return out
+        return impl
+
+
+@overload_method(types.NumPyRandomGeneratorType, 'logseries')
+def NumPyRandomGeneratorType_logseries(inst, p, size=None):
+    check_types(p, [types.Float, float], 'p')
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:876]: Same comment RE accepting `array_like` of floats and things that will cast to `float`.

File: numba/np/random/generator_methods.py [hunk 8]
```diff
@@ -819,3 +820,75 @@ def impl(inst, n, p , size=None):
...
+                raise ValueError("p < 0, p > 1 or p is NaN")
+            return np.int64(random_logseries(inst.bit_generator, p))
+        return impl
+    else:
+        check_size(size)
+
+        def impl(inst, p, size=None):
+            out = np.empty(size, dtype=np.int64)
+            for i in np.ndindex(size):
+                out[i] = random_logseries(inst.bit_generator, p)
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:892]: NumPy claims that `p` must be in range (0, 1) so Numba should ensure the same.

File: numba/np/random/distributions.py [hunk 9]
```diff
@@ -529,3 +529,48 @@ def random_poisson(bitgen, lam):
...
+
+@register_jitable
+def random_noncentral_f(bitgen, dfnum, dfden, nonc):
+    t = random_noncentral_chisquare(bitgen, dfnum, nonc) * dfden
+    return t / (random_chisquare(bitgen, dfden) * dfnum)
+
+
+@register_jitable
+def random_logseries(bitgen, p):
+    r = np_log1p(-p)
```
Comment (stuartarchibald) [inline numba/np/random/distributions.py:559]: If `p==1.0` Numba raises a `ZeroDivisionError` (the `error_model` needs setting to `numpy` in the decorator). 

NumPy just hangs, I suspect this is a bug as it ends up as a call to  `log1p(-p)` i.e. `log1p(-1)`, the result of which is `-HUGE_VAL` use of which in following operations ends up as `-inf`, this results in the `while` loop never exiting unless a call to `next_double` just so happens to hit `1.0` such that `(V >= p)` is true. What do you think?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. I've given this an initial review, the implementation largely looks good, but a bit more work is needed on matching NumPy with respect to types and error handling. There's also a potential NumPy bug in `logseries` that I've highlighted inline. Thanks again.

File: numba/np/random/generator_methods.py (hunk 1 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:842]: Good idea, I think this should be a generalised (i.e. used in rest of existing distributions as well). I'll add it as a task in the `Generator`s meta-issue.

File: docs/source/reference/numpysupported.rst [hunk 10]
```diff
@@ -627,7 +627,13 @@ The following :py:class:`Generator` methods are supported:
 * :func:`numpy.random.Generator().laplace()` (*)
 * :func:`numpy.random.Generator().logistic()` (*)
 * :func:`numpy.random.Generator().lognormal()` (*)
+* :func:`numpy.random.Generator().logseries()` (Accepts float values as well as datatypes
+ that cast to floats. Array values for p are currently not supported.)
 * :func:`numpy.random.Generator().negative_binomial()` (*)
+* :func:`numpy.random.Generator().noncentral_chisquare()` (Accepts float values as well as datatypes
+ that cast to floats. Array values for dfnum and nonc are currently not supported.)
+* :func:`numpy.random.Generator().noncentral_f()` (Accepts float values as well as datatypes
+ that cast to floats. Array values for dfnum, dfden and nonc are currently not supported.)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:636]:
```suggestion
* :func:`numpy.random.Generator().logseries()` (Accepts float values as well as datatypes
  that cast to floats. Array values for p are currently not supported.)
* :func:`numpy.random.Generator().negative_binomial()` (*)
* :func:`numpy.random.Generator().noncentral_chisquare()` (Accepts float values as well as datatypes
  that cast to floats. Array values for dfnum and nonc are currently not supported.)
* :func:`numpy.random.Generator().noncentral_f()` (Accepts float values as well as datatypes
  that cast to floats. Array values for dfnum, dfden and nonc are currently not supported.)
```
I think something like this ought to fix the doc build. IIRC continued item lines in bulleted lists need the same indent level as the start of the item.

File: numba/np/random/generator_methods.py (hunk 2 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:825]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 3 above)
Comment (kc611) [reply numba/np/random/generator_methods.py]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 4 above)
Comment (kc611) [reply numba/np/random/generator_methods.py]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 5 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:831]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 6 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:864]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 8 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:912]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 7 above)
Comment (kc611) [reply numba/np/random/generator_methods.py]: Resolved in [9bcf3e2](https://github.com/numba/numba/pull/8520/commits/9bcf3e2edbdb4f1fdb5417ccada72d589b6da88e)

File: numba/np/random/generator_methods.py (hunk 1 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:850]: Added as a task in #8519

File: docs/source/reference/numpysupported.rst [hunk 11]
```diff
@@ -627,7 +627,13 @@ The following :py:class:`Generator` methods are supported:
 * :func:`numpy.random.Generator().laplace()` (*)
 * :func:`numpy.random.Generator().logistic()` (*)
 * :func:`numpy.random.Generator().lognormal()` (*)
+* :func:`numpy.random.Generator().logseries()` (Accepts float values as well as datatypes
+  that cast to floats. Array values for p are currently not supported.)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:631]:
```suggestion
  that cast to floats. Array values for ``p`` are currently not supported.)
```
and line wrap to 80 chars as needed.

File: docs/source/reference/numpysupported.rst [hunk 12]
```diff
@@ -627,7 +627,13 @@ The following :py:class:`Generator` methods are supported:
...
 * :func:`numpy.random.Generator().negative_binomial()` (*)
+* :func:`numpy.random.Generator().noncentral_chisquare()` (Accepts float values as well as datatypes
+  that cast to floats. Array values for dfnum and nonc are currently not supported.)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:634]:
```suggestion
  that cast to floats. Array values for ``dfnum`` and ``nonc`` are currently not supported.)
```
and line wrap to 80 chars as needed.

File: docs/source/reference/numpysupported.rst [hunk 13]
```diff
@@ -627,7 +627,13 @@ The following :py:class:`Generator` methods are supported:
 * :func:`numpy.random.Generator().laplace()` (*)
 * :func:`numpy.random.Generator().logistic()` (*)
 * :func:`numpy.random.Generator().lognormal()` (*)
+* :func:`numpy.random.Generator().logseries()` (Accepts float values as well as datatypes
+  that cast to floats. Array values for p are currently not supported.)
 * :func:`numpy.random.Generator().negative_binomial()` (*)
+* :func:`numpy.random.Generator().noncentral_chisquare()` (Accepts float values as well as datatypes
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:633]: line wrap to 80 chars?

File: docs/source/reference/numpysupported.rst [hunk 14]
```diff
@@ -627,7 +627,13 @@ The following :py:class:`Generator` methods are supported:
...
+  that cast to floats. Array values for dfnum and nonc are currently not supported.)
+* :func:`numpy.random.Generator().noncentral_f()` (Accepts float values as well as datatypes
+  that cast to floats. Array values for dfnum, dfden and nonc are currently not supported.)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:636]:
```suggestion
  that cast to floats. Array values for ``dfnum``, ``dfden`` and ``nonc`` are currently not supported.)
```
and line wrap to 80 chars as needed.

File: numba/np/random/generator_methods.py [hunk 15]
```diff
@@ -819,3 +820,95 @@ def impl(inst, n, p , size=None):
...
+    check_types(nonc, [types.Float, types.Integer, int, float], 'nonc')
+    if isinstance(size, types.Omitted):
+        size = size.value
+
+    @register_jitable
+    def check_arg_bounds(dfnum, dfden, nonc):
+        if dfnum <= 0:
+            raise ValueError("df <= 0")
+        if dfden <= 0:
+            raise ValueError("df <= 0")
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:870]:
```suggestion
        if dfnum <= 0:
            raise ValueError("dfnum <= 0")
        if dfden <= 0:
            raise ValueError("dfden <= 0")
```
that testing didn't pick this up suggests that the tests are not sufficiently specific or this is not being tested.

File: numba/np/random/generator_methods.py [hunk 16]
```diff
@@ -819,3 +820,95 @@ def impl(inst, n, p , size=None):
...
+@overload_method(types.NumPyRandomGeneratorType, 'logseries')
+def NumPyRandomGeneratorType_logseries(inst, p, size=None):
+    check_types(p, [types.Float, types.Integer, int, float], 'p')
+    if isinstance(size, types.Omitted):
+        size = size.value
+
+    if is_nonelike(size):
+        def impl(inst, p, size=None):
+            if p < 0 or p > 1 or np.isnan(p):
+                raise ValueError("p < 0, p > 1 or p is NaN")
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:901]:
```suggestion
            if p < 0 or p >= 1 or np.isnan(p):
                raise ValueError("p < 0, p >= 1 or p is NaN")
```
xref: https://github.com/numpy/numpy/pull/22450

File: numba/np/random/generator_methods.py [hunk 17]
```diff
@@ -819,3 +820,95 @@ def impl(inst, n, p , size=None):
...
+
+@overload_method(types.NumPyRandomGeneratorType, 'noncentral_chisquare')
+def NumPyRandomGeneratorType_noncentral_chisquare(inst, df, nonc, size=None):
+    check_types(df, [types.Float, types.Integer, int, float], 'df')
+    check_types(nonc, [types.Float, types.Integer, int, float], 'nonc')
+    if isinstance(size, types.Omitted):
+        size = size.value
+
+    @register_jitable
+    def check_arg_bounds(df, nonc):
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:833]: Are these runtime exceptions being tested in the unit tests?

File: numba/np/random/generator_methods.py [hunk 18]
```diff
@@ -819,3 +820,95 @@ def impl(inst, n, p , size=None):
...
+            if p < 0 or p > 1 or np.isnan(p):
+                raise ValueError("p < 0, p > 1 or p is NaN")
+            return np.int64(random_logseries(inst.bit_generator, p))
+        return impl
+    else:
+        check_size(size)
+
+        def impl(inst, p, size=None):
+            if p < 0 or p > 1 or np.isnan(p):
+                raise ValueError("p < 0, p > 1 or p is NaN")
```
Comment (stuartarchibald) [inline numba/np/random/generator_methods.py:909]: Is this exception case being tested in the unit tests?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates. There's a few minor things to resolve predominantly RE testing of runtime errors, otherwise I think this is close to finished.

File: docs/source/reference/numpysupported.rst [hunk 19]
```diff
@@ -627,7 +627,16 @@ The following :py:class:`Generator` methods are supported:
 * :func:`numpy.random.Generator().laplace()` (*)
 * :func:`numpy.random.Generator().logistic()` (*)
 * :func:`numpy.random.Generator().lognormal()` (*)
+* :func:`numpy.random.Generator().logseries()` (Accepts float values as well as
+  datatypes that cast to floats. Array values for ``p`` are currently not
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:631]:
```suggestion
  data types that cast to floats. Array values for ``p`` are currently not
```

File: docs/source/reference/numpysupported.rst [hunk 20]
```diff
@@ -627,7 +627,16 @@ The following :py:class:`Generator` methods are supported:
...
+  supported.)
 * :func:`numpy.random.Generator().negative_binomial()` (*)
+* :func:`numpy.random.Generator().noncentral_chisquare()` (Accepts float values
+  as well as datatypes that cast to floats. Array values for ``dfnum`` and
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:635]:
```suggestion
  as well as data types that cast to floats. Array values for ``dfnum`` and
```

File: docs/source/reference/numpysupported.rst [hunk 21]
```diff
@@ -627,7 +627,16 @@ The following :py:class:`Generator` methods are supported:
...
+  ``nonc`` are currently not supported.)
+* :func:`numpy.random.Generator().noncentral_f()` (Accepts float values as well
+  as datatypes that cast to floats. Array values for ``dfnum``, ``dfden`` and
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:638]:
```suggestion
  as data types that cast to floats. Array values for ``dfnum``, ``dfden`` and
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for addressing the review comments @kc611, couple of minor typo things else looks good.

File: numba/np/random/generator_methods.py (hunk 18 above)
Comment (kc611) [reply numba/np/random/generator_methods.py]: Resolved in 280f269

File: numba/np/random/generator_methods.py (hunk 17 above)
Comment (kc611) [reply numba/np/random/generator_methods.py:833]: Resolved in 280f269

File: docs/source/reference/numpysupported.rst (hunk 13 above)
Comment (kc611) [reply docs/source/reference/numpysupported.rst]: Resolved in 280f269

Comment (kc611) [reply docs/source/reference/numpysupported.rst]: Resolved in 280f269

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes!

**Pull request merged**